extensions/amp-social-share/0.2/social-share.js

```diff
@@ -0,0 +1,188 @@
+      setHref(href.replace('?', '?&'));
+    }
+  });
+
+  /**
+   * Handle key presses on the element.
+   * @param {!Event} event
+   * @private
+   */
+  function handleKeyPress(event) {
+    const {key} = event;
+    if (key == Keys.SPACE || key == Keys.ENTER) {
+      event.preventDefault();
+      handleActivation();
+    }
+  }
+
+  /**
+   * Handle clicks on the element.
+   * @private
+   */
+  function handleClick() {
+    handleActivation();
+  }
+
+  /** @private */
+  function handleActivation() {
+    userAssert(href && target, 'Clicked before href is set.');
+    dev().assertString(href);
+    dev().assertString(target);
+    if (shareEndpoint === 'navigator-share:') {
+      devAssert(navigator.share !== undefined, 'navigator.share disappeared.');
+      // navigator.share() fails 'gulp check-types' validation on Travis
+      navigator['share'](parseQueryString(href.substr(href.indexOf('?'))));
+    } else {
+      const windowFeatures = 'resizable,scrollbars,width=640,height=480';
+      openWindowDialog(window, href, target, windowFeatures);
+    }
+  }
+
+  const attrs = {
+    onClick: handleClick,
+    onKeydown: handleKeyPress,
+    className: `amp-social-share-${props['type']}`,
+    style: socialShareStyle + props['style'],
+    role: 'button',
+    tabindex: props['tabindex'] || '0',
+  };
+
+  const render = props['render'];
+  useResourcesNotify();
+  return render(null, props['children'], createElement('div', attrs));
+}
+
+/**
+ * @param {string=} type
+ * @param {!../../../src/service/viewer-interface.ViewerInterface} viewer
+ * @param {!../../../src/service/platform-impl.Platform} platform
+ * @return {dict=}
+ */
+function getTypeConfigOrUndefined(type, viewer, platform) {
+  userAssert(type, 'The type property is required. %s');
+  userAssert(
+    !/\s/.test(type),
+    'Space characters are not allowed in the type property. %s'
+  );
+  if (type === 'system') {
+    // // navigator.share unavailable
+    if (!systemShareSupported(viewer, platform)) {
+      return;
+    }
+  } else {
+    // system share wants to be unique
+    const systemOnly =
+      systemShareSupported(viewer, platform) &&
+      !!window.document.querySelectorAll(
+        'amp-social-share[type=system][data-mode=replace]'
+      ).length;
+    if (systemOnly) {
+      return;
+    }
+  }
+  const typeConfig = getSocialConfig(type) || dict();
+  if (typeConfig['obsolete']) {
+    user().warn(`Skipping obsolete share button ${type}`);
+    return;
+  }
+  return typeConfig;
+}
+
+/**
+ * @param {!../../../src/service/viewer-interface.ViewerInterface} viewer
+ * @param {!../../../src/service/platform-impl.Platform} platform
+ * @return {boolean}
+ */
+function systemShareSupported(viewer, platform) {
+  // Chrome exports navigator.share in WebView but does not implement it.
+  // See https://bugs.chromium.org/p/chromium/issues/detail?id=765923
+  const isChromeWebview = viewer.isWebviewEmbedded() && platform.isChrome();
```

Comment:
I think both of these should be done in `init`, and it should throw if it's improperly configured. I think these are both highly specific to AMP (at least, the services are). Do we need to support this behavior in all Bento Modes?